diff --git a/.gitignore b/.gitignore
index bc9164db98a3..9c292436a98e 100644
--- a/.gitignore
+++ b/.gitignore
@@ -9,6 +9,17 @@
 #
 # Normal rules (sorted alphabetically)
 #
+config.muk
+config-4.14*
+config-5.4*
+okvmstat
+*.diff
+*.sav
+vmioctl
+*.orig
+*.rej
+*.old
+*.prev
 .*
 *.a
 *.asn1.[ch]
diff --git a/Makefile b/Makefile
index 96bf8219a73d..584b931b8950 100644
--- a/Makefile
+++ b/Makefile
@@ -459,6 +459,8 @@ KBUILD_CFLAGS   := -Wall -Wundef -Werror=strict-prototypes -Wno-trigraphs \
 		   -fno-strict-aliasing -fno-common -fshort-wchar -fno-PIE \
 		   -Werror=implicit-function-declaration -Werror=implicit-int \
 		   -Wno-format-security \
+		   -fmax-errors=4 -Werror=return-type -Werror=int-conversion \
+		   -Werror=misleading-indentation       \
 		   -std=gnu89
 KBUILD_CPPFLAGS := -D__KERNEL__
 KBUILD_AFLAGS_KERNEL :=
@@ -917,6 +919,9 @@ KBUILD_CPPFLAGS += $(KCPPFLAGS)
 KBUILD_AFLAGS   += $(KAFLAGS)
 KBUILD_CFLAGS   += $(KCFLAGS)
 
+#Mukesh: Add ours after all flgs are set, so our setting is at the end
+KBUILD_CFLAGS   += -Wno-error=date-time
+
 KBUILD_LDFLAGS_MODULE += --build-id
 LDFLAGS_vmlinux += --build-id
 
