diff --git a/.gitignore b/.gitignore
index 689b85dd3e16..edb36a7ea26d 100644
--- a/.gitignore
+++ b/.gitignore
@@ -9,6 +9,8 @@
 #
 # Normal rules (sorted alphabetically)
 #
+*.orig
+*.rej
 .*
 *.a
 *.bin
diff --git a/Makefile b/Makefile
index 58370018560f..2825dd8c1fd0 100644
--- a/Makefile
+++ b/Makefile
@@ -850,6 +850,12 @@ KBUILD_CPPFLAGS += $(ARCH_CPPFLAGS) $(KCPPFLAGS)
 KBUILD_AFLAGS   += $(ARCH_AFLAGS)   $(KAFLAGS)
 KBUILD_CFLAGS   += $(ARCH_CFLAGS)   $(KCFLAGS)
 
+#Mukesh: Add ours after all flgs are set, so our setting is at the end
+KBUILD_CFLAGS   += -Wno-error=date-time -Wno-attribute-alias \
+		   -fmax-errors=4 -Werror=return-type -Werror=int-conversion \
+		   -Wno-stringop-truncation -Werror=misleading-indentation \
+		   -Wno-stringop-overflow -Wno-packed-not-aligned
+
 # Use --build-id when available.
 LDFLAGS_BUILD_ID := $(patsubst -Wl$(comma)%,%,\
 			      $(call cc-ldoption, -Wl$(comma)--build-id,))
diff --git a/arch/x86/kernel/cpu/microcode/Makefile b/arch/x86/kernel/cpu/microcode/Makefile
index e5a819da5c6a..46e9645046ae 100644
--- a/arch/x86/kernel/cpu/microcode/Makefile
+++ b/arch/x86/kernel/cpu/microcode/Makefile
@@ -1,3 +1,4 @@
+KBUILD_CFLAGS   += -Wno-incompatible-pointer-types
 microcode-y				:= core.o
 obj-$(CONFIG_MICROCODE)			+= microcode.o
 microcode-$(CONFIG_MICROCODE_INTEL)	+= intel.o
diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 42ceb26b069c..af881da5fd04 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -4881,7 +4881,7 @@ int handle_ud(struct kvm_vcpu *vcpu)
 	struct x86_exception e;
 
 	if (force_emulation_prefix &&
-	    kvm_read_guest_virt(&vcpu->arch.emulate_ctxt,
+	    kvm_read_guest_virt(vcpu,
 				kvm_get_linear_rip(vcpu), sig, sizeof(sig), &e) == 0 &&
 	    memcmp(sig, "\xf\xbkvm", sizeof(sig)) == 0) {
 		kvm_rip_write(vcpu, kvm_rip_read(vcpu) + sizeof(sig));
diff --git a/drivers/block/oracleasm/Makefile b/drivers/block/oracleasm/Makefile
index 8fafb5323035..bb1c7df9eefa 100644
--- a/drivers/block/oracleasm/Makefile
+++ b/drivers/block/oracleasm/Makefile
@@ -3,4 +3,5 @@ obj-$(CONFIG_ORACLEASM) := oracleasm.o
 oracleasm-y += driver.o transaction_file.o
 oracleasm-$(CONFIG_BLK_DEV_INTEGRITY) += integrity.o
 CFLAGS_driver.o = -I$(src)
+ccflags-y += -Wno-designated-init
 
diff --git a/drivers/vhost/Kconfig b/drivers/vhost/Kconfig
index cfdecea5078f..b580885243f7 100644
--- a/drivers/vhost/Kconfig
+++ b/drivers/vhost/Kconfig
@@ -12,7 +12,7 @@ config VHOST_NET
 
 config VHOST_SCSI
 	tristate "VHOST_SCSI TCM fabric driver"
-	depends on TARGET_CORE && EVENTFD && m
+	depends on TARGET_CORE && EVENTFD
 	select VHOST
 	default n
 	---help---
diff --git a/drivers/virtio/virtio_pci_common.h b/drivers/virtio/virtio_pci_common.h
index 135ee3cf7175..b82ef625d359 100644
--- a/drivers/virtio/virtio_pci_common.h
+++ b/drivers/virtio/virtio_pci_common.h
@@ -118,7 +118,7 @@ enum {
 };
 
 /* Convert a generic virtio device to our structure */
-static struct virtio_pci_device *to_vp_device(struct virtio_device *vdev)
+static inline struct virtio_pci_device *to_vp_device(struct virtio_device *vdev)
 {
 	return container_of(vdev, struct virtio_pci_device, vdev);
 }
